### PR TITLE
Feature/build docker images

### PR DESCRIPTION
**Pull request by lwronski**
Building docker images on CI and used them for testing.

### CODE REVIEW

File: .github/scripts/docker/DockerFile [hunk 1]
```diff
@@ -0,0 +1,3 @@
+FROM gcr.io/distroless/java-debian10
```
Comment (alexarchambault) [inline .github/scripts/docker/DockerFile]: I think `gcr.io/distroless/base-debian10` should work as a base, doesn't it?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -10,6 +10,7 @@ on:
 jobs:
   jvm-tests:
     runs-on: ${{ matrix.OS }}
+    if: 1 == 2
```
Comment (alexarchambault) [inline .github/workflows/ci.yml]: Remove this?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -45,6 +46,7 @@ jobs:
 
   native-tests:
     runs-on: ${{ matrix.OS }}
+    if: 2 == 3
```
Comment (alexarchambault) [inline .github/workflows/ci.yml]: Remove this too?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -141,6 +148,7 @@ jobs:
 
   reference-doc:
     runs-on: ubuntu-latest
+    if: 2 == 1
```
Comment (alexarchambault) [inline .github/workflows/ci.yml]: Remove this?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -101,6 +107,7 @@ jobs:
 
   native-static-tests:
     runs-on: ubuntu-latest
+    if: 2 == 1
```
Comment (alexarchambault) [inline .github/workflows/ci.yml]: Remove this?

File: build.sc [hunk 6]
```diff
@@ -357,6 +367,11 @@ trait JvmIntegrationCore extends CliIntegrationCore {
   def cliKind = "jvm"
 }
 
+trait DockerIntegrationCore extends CliIntegrationCore {
+  def cliKind = "docker"
+  def testLauncher = `cli-core`.launcher()
```
Comment (alexarchambault) [inline build.sc]: Do we need to pass a test launcher to the docker tests?

File: modules/integration-core/src/test/scala/scala/cli/integration/RunDockerTests.scala [hunk 7]
```diff
@@ -0,0 +1,45 @@
+package scala.cli.integration
+
+import com.eed3si9n.expecty.Expecty.expect
+
+import java.io.{ByteArrayOutputStream, File}
+import java.nio.charset.Charset
+
+import scala.util.Properties
+
+class RunDockerTests extends munit.FunSuite {
```
Comment (alexarchambault) [inline modules/integration-core/src/test/scala/scala/cli/integration/RunDockerTests.scala]: This file has 4 space indent, right? We use 2 space indent most of the time.

File: modules/integration-core/src/test/scala/scala/cli/integration/RunDockerTests.scala [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+                System.err.write(b, 0, len)
+                },
+                mergeErrIntoOut = true
+            )
+            val output = new String(rawOutput.toByteArray, Charset.defaultCharset())
+            expect(output.linesIterator.toVector.last == message)
+        }
+    }
+
+    if (Properties.isLinux && TestUtil.cliKind == "docker")
```
Comment (alexarchambault) [inline modules/integration-core/src/test/scala/scala/cli/integration/RunDockerTests.scala]: We shouldn't need this check… We'll only run these tests on systems that support it. (Incidentally, running them on macOS or Windows might work, as the mostly static binary is built from docker too.)

File: .github/scripts/docker/DockerFile (hunk 1 above)
Comment (lwronski) [reply .github/scripts/docker/DockerFile]: If I use `gcr.io/distroless/base-debian10`, the test is fail -
` /usr/local/bin/scala: error while loading shared libraries: libstdc++.so.6: cannot open shared object file: No such file or directory`

File: build.sc (hunk 6 above)
Comment (lwronski) [reply build.sc]: Probably not, so how should I implement method testLauncher in DockerIntegrationCore?

File: .github/scripts/docker/DockerFile (hunk 1 above)
Comment (alexarchambault) [reply .github/scripts/docker/DockerFile]: The mostly-static binary is precisely made not to depend on `libstdc++`. It works fine [here](https://github.com/VirtuslabRnD/scala-cli/blob/c01a212af47e61a2c530f7bf0b4186c4a58bf991/modules/integration-core/src/test/scala/scala/cli/integration/RunTestDefinitions.scala#L809-L811).

File: .github/workflows/ci.yml (hunk 2 above)
Comment (alexarchambault) [reply .github/workflows/ci.yml]: This still appear in the first commit. Same for the other `if`s.

File: build.sc (hunk 6 above)
Comment (alexarchambault) [reply build.sc]: Just don't use `CliIntegrationBase` in `DockerIntegrationCore`.

File: modules/integration-core/src/test/scala/scala/cli/integration/RunDockerTests.scala (hunk 7 above)
Comment (alexarchambault) [reply modules/integration-core/src/test/scala/scala/cli/integration/RunDockerTests.scala]: I still see 4 space indents in the first commit.

File: modules/integration-core/src/test/scala/scala/cli/integration/RunDockerTests.scala [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+import com.eed3si9n.expecty.Expecty.expect
+
+import java.io.{ByteArrayOutputStream, File}
+import java.nio.charset.Charset
+
+import scala.util.Properties
+
+class RunDockerTests extends munit.FunSuite {
+
+    def simplRunAppDocker(): Unit = {
```
Comment (alexarchambault) [inline modules/integration-core/src/test/scala/scala/cli/integration/RunDockerTests.scala]: `s/simplRunAppDocker/simpleRunAppDocker/g`?

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -92,7 +92,7 @@ jobs:
     - name: Generate docker image
       run: .github/scripts/generate-docker-images-ubuntu.sh
     - name: Docker integration tests
-      run: ./mill integration.docker.test 'scala.cli.integration.RunDockerTests.*'
+      run:  ./mill integration.docker.test
```
Comment (alexarchambault) [inline .github/workflows/ci.yml]: Unnecessary space before `./mill`

File: .github/workflows/ci.yml (hunk 2 above)
Comment (lwronski) [reply .github/workflows/ci.yml]: I have just squashed all commits and remove this change from history

File: .github/scripts/docker/DockerFile (hunk 1 above)
Comment (lwronski) [reply .github/scripts/docker/DockerFile]: It's works. Locally, I was using different binary with scala-cli so it throws an error.

File: build.sc [hunk 11]
```diff
@@ -53,7 +53,22 @@ object `integration-core` extends Module {
   }
 }
 
+object `integration-docker` extends CliIntegrationDockerCore {
```
Comment (alexarchambault) [inline build.sc]: Do we really need that one? I think by renaming `modules/integration-docker` to `modules/integration/docker`, `integration.docker` below should automatically pick the right sources.

Comment (lwronski) [reply build.sc]: Yes, we don't need this definition of module. I moved `modules/integration-docker` to `modules/integration/docker` and it correct pick the right sources.

**Review by alexarchambault — APPROVED** [review]

**Pull request merged**